### ISSUE TITLE
Volume_15min is creating unnecessary 0 volume bins at 3 legged intersections

### ISSUE THREAD

**Issue by rickl4** [CityofToronto/bdit_data-sources#144]
At 3 legged intersections, specifically Adelaide/Bathurst and Front/Bathurst, W leg EB bins are being created for vehicles. While this doesn't affect the methodology of the process, it does create extra data. A similar table to `miovision.intersection_movements` should be created.

**Comment by rickl4** [comment]
There are more invalid movements in the data than I originally thought. [This query](https://github.com/CityofToronto/bdit_data-sources/blob/automate_miovision_aggregation/volumes/miovision/sql/select-invalid_movements_tmc.sql) selects all counts with invalid movements in `volumes`, while this query does the same for `volumes_15min`.

```SQL
SELECT intersection_uid, datetime_bin::Date, count(*) FROM miovision.volumes_15min A

WHERE (intersection_uid IN (1,2,3,4,5,26)
AND leg='W' 
AND dir='WB'
AND classification_uid IN (1,3,4,5))

OR (intersection_uid IN (26,27,28,29,30,31)
AND leg='E' 
AND dir='EB'
AND classification_uid IN (1,3,4,5))

OR (intersection_uid IN (14)
AND leg='N' 
AND dir='NB'
AND classification_uid IN (1,3,4,5))

OR (intersection_uid IN (16)
AND leg='S' 
AND dir='SB'
AND classification_uid IN (1,3,4,5))
```
There are 132 turning movements in `intersection_movements` that are not valid, and approximately 4000 entries in `miovision_raw_data` that are not valid movements. The largest portion of these entries was Adelaide/Bathurst on April 7th.

**Comment by rickl4** [comment]
I created `rliu.intersection_movements` which only has valid movements. Replacing the data in `miovision.intersection_movements` with the data in this table would remove invalid movements in `volumes_15min_tmc`. 

In `aggregate_15min`, this query should be used:

```SQL
WITH transformed AS (
	SELECT 	A.intersection_uid,
		A.datetime_bin,
		A.classification_uid,
		B.leg_new as leg,
		B.dir,
		SUM(A.volume) AS volume,
		array_agg(volume_15min_tmc_uid) as uids

	FROM rliu.volumes_15min_tmc A
	INNER JOIN 
		(SELECT intersection_uid, classification_uid, leg_old, A.movement_uid, leg_new, dir FROM miovision.movement_map A
		 JOIN rliu.intersection_movements B ON B.movement_uid=A.movement_uid AND B.leg=A.leg_old) B 
	ON B.leg_old = A.leg AND B.movement_uid = A.movement_uid AND B.intersection_uid=A.intersection_uid AND B.classification_uid=A.classification_uid
	GROUP BY A.intersection_uid, A.datetime_bin, A.classification_uid, B.leg_new, B.dir
	ORDER BY A.datetime_bin, A.intersection_uid, A.classification_uid, B.leg_new, B.dir
)

INSERT INTO rliu.volumes_15min(intersection_uid, datetime_bin, classification_uid, leg, dir, volume)
SELECT A.intersection_uid, B.datetime_bin, A.classification_uid, A.leg, A.dir, COALESCE(C.volume,-1) AS volume
FROM (SELECT intersection_uid, classification_uid, leg, dir FROM transformed GROUP BY intersection_uid, classification_uid, leg, dir) AS A
INNER JOIN (SELECT intersection_uid, datetime_bin FROM transformed GROUP BY intersection_uid, datetime_bin) AS B USING (intersection_uid)
LEFT JOIN transformed C USING (intersection_uid, datetime_bin, classification_uid, leg, dir);
```

Inner joining `intersection_movements` onto `movement_map` will limit the query to only valid movements.

**Comment by radumas** [comment]
Sounds good. I think filtering out invalid movements after the 1-minute volumes seems like a good point in the processing so we can QC on whether a large number of invalid movements have appeared in the data.

**Comment by rickl4** [comment]
I developed a [function](https://github.com/CityofToronto/bdit_data-sources/blob/automate_miovision_aggregation/volumes/miovision/sql/function-find_invalid_movements.sql) that warns the user if there are more than 1000 invalid movements during the past 30 days. If this occurs, then a QC check should be done. I chose 1000 since Front/Bathurst had 1700 invalid movements during April. 

The threshold and date range can be changed. We can also integrate it into `aggregate_15_min_tmc` or as a trigger, and change the `RAISE NOTICE` to `RAISE EXCEPTION` if we want a more automatic way of checking invalid movements.

**Comment by radumas** [comment]
Good start and nice thinking about having a separate function that can be integrated into other functions.

A thought: I think it would be easier to maintain (and extend) a check against movements which are explicitly allowed in `intersection_movements` rather than enumerating all possible wrong movements. 

A further thought: this could be integrated into the API pipeline (if we want?). [`psycopg2` can capture the notices raised](http://initd.org/psycopg/docs/connection.html#connection.notices) during a query.

**Comment by radumas** [comment]
I'm also going to delete movements from `intersection_movements` that match the logic @rickl4 has in [this function](https://github.com/CityofToronto/bdit_data-sources/blob/automate_miovision_aggregation/volumes/miovision/sql/function-find_invalid_movements.sql#L15-L65) and then can delete from the aggregate tables following that.

**Comment by rickl4** [comment]
Checking the API, Richmond/Spadina, Adelaide Spadina, and Adelaide Jarvis all have daily instances where the number of invalid counts are over 5000 until May 30. Other intersections had a few dates with high numbers of invalid movements, but these three intersections had counts over 5000 every day until May 30th. 

However after May 30, there were no invalid movements anywhere during any 1 minute bin.

**Comment by radumas** [comment]
I'm going to spin off checking this for the API into a new task so we can close this off once the function for checking data from data dumps is finished.

**Comment by radumas** [comment]
Hey @rickl4, which of the movements in [that function](https://github.com/CityofToronto/bdit_data-sources/blob/automate_miovision_aggregation/volumes/miovision/sql/function-find_invalid_movements.sql#L15-L65) are illegal vs. illogical? I think we should still include things like illegal left turns, but flag vehicles going the wrong way on one-way streets or where there is no approach, such as the W leg of Bathurst.

**Comment by radumas** [comment]
Deleted data from `miovision.volumes_15min_tmc` and `miovision.intersection_movements`:
```sql
WHERE (intersection_uid IN (2,3,4)
		AND leg='E' 
		AND classification_uid IN (1,3,4,5))

		OR (intersection_uid IN (27,28,29,31,1,5)
		AND leg='W' 
		AND classification_uid IN (1,3,4,5))

		OR (intersection_uid IN (14)
		AND leg='S' 
		AND classification_uid IN (1,3,4,5))

		OR (intersection_uid IN (16)
		AND leg='N' 
		AND classification_uid IN (1,3,4,5))

		OR (intersection_uid IN (26)
		AND classification_uid IN (1,3,4,5)
		AND ((leg IN ('E','W') 
		AND movement_uid IN (1,4))
		OR (leg IN ('N','S') 
		AND movement_uid IN (2,3))))

		OR (intersection_uid IN (30)
		AND leg IN ('W') 
		AND classification_uid IN (1,3,4,5)
		AND movement_uid IN (1,4))

		OR (intersection_uid IN (27,28,29,31,30)
		AND classification_uid IN (1,3,4,5)
		AND ((leg IN ('S') 
		AND movement_uid IN (3)) OR (leg IN ('N') 
		AND movement_uid IN (2))))

		OR (intersection_uid IN (2,3,4)
		AND classification_uid IN (1,3,4,5)
		AND ((leg IN ('S') 
		AND movement_uid IN (2)) OR (leg IN ('N') 
		AND movement_uid IN (3))))

		OR (intersection_uid IN (14)
		AND classification_uid IN (1,3,4,5)
		AND ((leg IN ('E') 
		AND movement_uid IN (3)) OR (leg IN ('W') 
		AND movement_uid IN (2))))

		OR (intersection_uid IN (16)
		AND classification_uid IN (1,3,4,5)
		AND ((leg IN ('E') 
		AND movement_uid IN (2)) OR (leg IN ('W') 
		AND movement_uid IN (3))))
```
